FreeRTOS/FreeRTOS-Plus-TCP
# IPv4/Single: Let send() stop blocking after a connection reset

Description
-----------
While `FreeRTOS_send()` is waiting for free space in the transmission buffer, it may happen that a disconnection finishes.
This was reported on the [FreeRTOS forum](https://forums.freertos.org/t/are-issues-from-sourceforge-still-relevant/15957) by [@diydsp](https://forums.freertos.org/u/diydsp).

Summary
---
~~~diff
BaseType_t FreeRTOS_send( Socket_t xSocket, ... )
{
    xByteCount = ( BaseType_t ) prvTCPSendCheck( pxSocket, uxDataLength );

    if( xByteCount > 0 )
    {
        /* Try sending till there is a timeout or all bytes have been sent. */
        while( !timed_out )
        {
            /* If txStream has space. */
            if( xByteCount > 0 )
            {
                /* Send data to the TCP TX stream. */
                xByteCount = uxStreamBufferAdd();
                xBytesLeft -= xByteCount;
                if( xBytesLeft == 0 )
                {
                    break;
                }
            }

            /* Go sleeping until there is new space or disconnected, or timeout */
            xEventGroupWaitBits( pxSocket->xEventGroup, eSOCKET_SEND | eSOCKET_CLOSED, ... );

+           xResult = ( BaseType_t ) prvTCPSendCheck( pxSocket, uxDataLength );
+
+           if( xResult < 0 )
+           {
+               /* In a meanwhile, the connection has dropped, stop iterating. */
+               break;
+           }
            xByteCount = ( BaseType_t ) uxStreamBufferGetSpace( pxSocket->u.xTCP.txStream );
        }
    }

    return xByteCount;
}
~~~

Test Steps
-----------
Let the DUT send loads of data, while the TCP connection gets disrupted from the other side. `FreeRTOS_send()` should stop blocking and return with the error `-pdFREERTOS_ERRNO_ENOTCONN`.

Related Issue
-----------
The earlier behaviour becomes a real problem when the time-out equals `portMAX_DELAY`. So for realistic testing, set the property `FREERTOS_SO_SNDTIMEO` to `portMAX_DELAY`.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
